### PR TITLE
Add boto3 extra_args field for upload operation

### PR DESCRIPTION
**Pull request by benbemoh**

### CODE REVIEW

File: cloudpathlib/s3/s3client.py [hunk 1]
```diff
@@ -65,6 +66,8 @@ def __init__(
                 Parameterize it to access a customly deployed S3-compatible object store such as MinIO, Ceph or any other.
             boto3_transfer_config (Optional[dict]): Instantiated TransferConfig for managing s3 transfers.
                 (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#boto3.s3.transfer.TransferConfig)
+            boto3_upload_extra_args (Optional[dict]): extra arguments for managing uploading objects s3.
```
Comment (pikulmar) [inline cloudpathlib/s3/s3client.py]: typo: "... _to S3_."

File: cloudpathlib/s3/s3client.py [hunk 2]
```diff
@@ -258,6 +262,8 @@ def _upload_file(self, local_path: Union[str, os.PathLike], cloud_path: S3Path)
         obj = self.s3.Object(cloud_path.bucket, cloud_path.key)
 
         extra_args = {}
```
Comment (pikulmar) [inline cloudpathlib/s3/s3client.py]: You can simplify this even further: `extra_args = self.boto3_upload_extra_args.copy()` (`__init__` ensures that it is a dict).

**Review by pikulmar — COMMENTED** [review]
Some cosmetics only.

File: cloudpathlib/s3/s3client.py (hunk 1 above)
Comment (pikulmar) [reply cloudpathlib/s3/s3client.py]: Addressed in 4f29d1f

File: cloudpathlib/s3/s3client.py (hunk 2 above)
Comment (benbemoh) [reply cloudpathlib/s3/s3client.py]: Addressed in [4f29d1f](https://github.com/TerrawattEnergy/cloudpathlib/pull/1/commits/4f29d1f3941ccb08fcbeed95eef2d36c1be37985)

File: cloudpathlib/s3/s3client.py [hunk 3]
```diff
@@ -99,7 +99,9 @@ def __init__(
             self.client = self.sess.client("s3", endpoint_url=endpoint_url)
 
         self.boto3_transfer_config = boto3_transfer_config
-        self.boto3_upload_extra_args: Final[Dict[str, str]] = boto3_upload_extra_args or {}
+        self.boto3_upload_extra_args: Final[Dict[str, str]] = (
+            boto3_upload_extra_args.copy() if boto3_upload_extra_args else {}
```
Comment (pikulmar) [inline cloudpathlib/s3/s3client.py]: Python golf suggestion: `(boto3_upload_extra_args or {}).copy()`

File: cloudpathlib/s3/s3client.py (hunk 2 above)
Comment (pikulmar) [reply cloudpathlib/s3/s3client.py]: Addressed in [4f29d1f](https://github.com/TerrawattEnergy/cloudpathlib/pull/1/commits/4f29d1f3941ccb08fcbeed95eef2d36c1be37985)

**Review by pikulmar — APPROVED** [review]
LGTM

File: cloudpathlib/s3/s3client.py (hunk 3 above)
Comment (benbemoh) [reply cloudpathlib/s3/s3client.py]: Addressed in [46a5655](https://github.com/TerrawattEnergy/cloudpathlib/pull/1/commits/46a56558e0fbcf9a0207bdd7fe9e91853ff57290)

**Review by pikulmar — APPROVED** [review]

**Pull request merged**